ratings/ratings.py

```diff
@@ -0,0 +1,298 @@
+    model_ids = set([d[0] for d in data]).union(set([d[1] for d in data]))
+
+    # Map model_ids to a contiguous range.
+    ordered = sorted(model_ids)
+    new_id = {}
+    for i, m in enumerate(ordered):
+        new_id[m] = i
+
+    # A function to rewrite the model_ids in our pairs
+    def ilsr_data(d):
+        p1, p2 = d
+        p1 = new_id[p1]
+        p2 = new_id[p2]
+        return (p1, p2)
+
+    pairs = list(map(ilsr_data, data))
+    ilsr_param = choix.ilsr_pairwise(
+        len(ordered),
+        pairs,
+        alpha=0.0001,
+        max_iter=800)
+
+    hessian = choix.opt.PairwiseFcts(pairs, penalty=.1).hessian(ilsr_param)
+    std_err = np.sqrt(np.diagonal(np.linalg.inv(hessian)))
+
+    # Elo conversion
+    elo_mult = 400 / math.log(10)
+
+    min_rating = min(ilsr_param)
+    ratings = {}
+
+    for model_id, param, err in zip(ordered, ilsr_param, std_err):
+        ratings[model_id] = (elo_mult * (param - min_rating), elo_mult * err)
+
+    return ratings
+
+
+def top_n(n=20):
+  rating_tups = compute_ratings()
+  r = compute_ratings()
+  return [(model_num_for(k),v) for v,k in
+          sorted([(v,k) for k,v in r.items()])[-n:][::-1]]
+
+def ingest_dirs(root, dirs):
+    for d in dirs:
+        if os.path.isdir(os.path.join(root,d)):
+            fs = [os.path.join(root, d, f) for f in os.listdir(os.path.join(root, d))]
+            print("Importing from {}".format(d))
+            import_files(fs)
+
+
+def last_timestamp():
+    db = sqlite3.connect("ratings.db")
+    with db:
+        ts = db.execute("select timestamp from games order by timestamp desc limit 1").fetchone()
+    return ts[0] if ts else None
+
+
+def suggest_pairs(top_n=10, per_n=3):
+    """ Find the maximally interesting pairs of players to match up
+    First, sort the ratings by uncertainty.
+    Then, take the ten highest players with the highest uncertainty
+    For each of them, call them `p1`
+    Sort all the models by their distance from p1's rating and take the 20 closest
+    Choose the model with the greatest age difference from among those candidates
+
+    'ratings' is a list of (model_num, rating, uncertainty) tuples
+
+    Returns a list of *model numbers*, not model ids.
+    """
+    db = sqlite3.connect("ratings.db")
+    data = db.execute("select model_winner, model_loser from wins").fetchall()
+    bucket_ids = [id[0] for id in db.execute("select id from models where bucket = ?", (fsdb.models_dir(),)).fetchall()]
+    bucket_ids.sort()
+    data = [d for d in data if d[0] in bucket_ids and d[1] in bucket_ids]
+
+    ratings = [(model_num_for(k), v[0], v[1]) for k,v in compute_ratings(data).items()]
+    ratings.sort()
+    ratings = ratings[70:] # filter off the first 100 models, which improve too fast.
+
+    ratings.sort(key=lambda r: r[2], reverse=True)
+
+    res = []
+    for p1 in ratings[:top_n]:
+        candidate_p2s = list(sorted(ratings, key=lambda p2_tup: abs(p1[1] - p2_tup[1])))[:20]
```

Comment:
no need for list()
does [1:20] removes the need for list compression below?

Reply:
yep, Done.